test/benchmarks/png_funcs.h

```diff
@@ -0,0 +1,150 @@
+            /* Pretty unlikely but we'll put it here just in case */
+            fprintf(stderr, "realloc failed, exiting\n");
+            exit(1);
+        }
+
+        dat->buf_rem += 2*len;
+    }
+
+    memcpy(dat->buf + dat->len, data, len);
+    dat->len = curSize;
+    dat->buf_rem -= len;
+}
+
+#if 0
+static void init_incompressible(png_bytep img_bytes, size_t num_pix) {
+    /* fix the random seed so that there's no variance for compressibility */    
+    srand(0);
+
+    for (size_t i = 0; i < 3*num_pix; ++i) {
+        img_bytes[i] = (png_byte)(rand() % 255);
+    }
+}
+#endif
+
+static void init_compressible(png_bytep buf, size_t num_pix) {
+    /* It doesn't actually matter what we make this, but for
+     * the sake of a reasonable test image, let's make this
+     * be a stripe of R, G, & B, with no alpha channel */
+    int i = 0;
+    int red_stop = num_pix/3;
+    int blue_stop = 2*num_pix/3;
+    int green_stop = num_pix;
+
+    for (; i < red_stop; ++i) {
+       buf[3*i] = 255; 
+       buf[3*i+1] = 0;
+       buf[3*i+2] = 0; 
+    }
+
+    for (; i < blue_stop; ++i) {
+       buf[3*i] = 0; 
+       buf[3*i+1] = 255; 
+       buf[3*i+2] = 0; 
+    }
+
+    for (; i < green_stop; ++i) {
+       buf[3*i] = 0; 
+       buf[3*i+1] = 0; 
+       buf[3*i+2] = 255; 
+    }
+}
+
+static inline void encode_png(png_bytep buf, png_dat *outpng, int comp_level) {
+        png_structp png = png_create_write_struct(PNG_LIBPNG_VER_STRING, NULL, NULL, NULL);
+
+        /* Most of this error handling is _likely_ not necessary. Likewise it's likely
+         * a lot of this stuff can be done in the setup function to avoid measuring this
+         * fixed setup time, but for now we'll do it here */
+        if (!png) abort();
+
+        png_infop info  = png_create_info_struct(png);
+        if (!info) abort();
+
+        png_set_write_fn(png, outpng, png_write_cb, NULL);
+        png_bytep png_row_ptrs[IMHEIGHT];
+        for (int i = 0; i < IMHEIGHT; ++i) {
+            png_row_ptrs[i] = (png_bytep)&buf[3*i*IMWIDTH];
+        }
+
+        png_set_IHDR(png, info, IMWIDTH, IMHEIGHT, 8, PNG_COLOR_TYPE_RGB,
+                     PNG_INTERLACE_NONE, PNG_COMPRESSION_TYPE_DEFAULT,
+                     PNG_FILTER_TYPE_DEFAULT);
+
+        png_write_info(png, info);
+        png_set_compression_level(png, comp_level);
+        png_set_filter(png, 0, PNG_FILTER_NONE);
+        png_write_image(png, (png_bytepp)png_row_ptrs);
+        png_write_end(png, NULL);
+        png_destroy_write_struct(&png, &info);
+}
+
+static void read_from_pngdat(png_structp png, png_bytep out, png_size_t bytes_to_read)
+{
+    png_parse_dat *io = (png_parse_dat*)png_get_io_ptr(png);
+    memcpy(out, io->cur_pos, bytes_to_read);
+    io->cur_pos += bytes_to_read;
+}
+
+static inline void decode_png(png_parse_dat *dat, png_bytep out_bytes) {
+    png_bytep out_rows[IMHEIGHT];
+    for (size_t i = 0; i < IMHEIGHT; ++i) out_rows[i] = &out_bytes[IMWIDTH*i];
```

Comment:
Break this `for` line into two lines. It is kind of harder to read it being on one line.